### PR TITLE
Representation of MHW in a frame

### PR DESCRIPTION
**Pull request by Kamil-Kisielewicz**

### CODE REVIEW

File: mhw_analysis/ai/ai_mhw_images.py [hunk 1]
```diff
@@ -0,0 +1,295 @@
...
+import os
+import numpy as np
+from matplotlib import pyplot as plt
+
+import iris
+
+import sys
+
+# create module paths
+MHW_PATH = '/Users/kamil/Desktop/research2020/mhw_analysis'
```
Comment (profxj) [inline mhw_analysis/ai/ai_mhw_images.py:12]: This is the Repository path, right?
You can access it with:

```
from pkg_resources import resource_filename
resource_filename('mhw_analysis', SUB_FOLDER)
```

where SUB_FOLDER is any sub-folder in the repo.

Please do that.

Same for `OCEANPY_PATH`.

No hard-coding paths.

File: mhw_analysis/ai/ai_mhw_images.py [hunk 2]
```diff
@@ -0,0 +1,295 @@
...
+OCEANPY_PATH = '/Users/kamil/Desktop/research2020/oceanpy'
+os.environ['NOAA_OI'] = '/Users/kamil/Desktop/NOAA-OI-SST-V2'
```
Comment (profxj) [inline mhw_analysis/ai/ai_mhw_images.py:14]: Don't put this in code.  Just define this in your `bash` profile.

File: mhw_analysis/ai/ai_mhw_images.py [hunk 3]
```diff
@@ -0,0 +1,295 @@
...
+
+# append paths
+sys.path.append(MHW_PATH)
```
Comment (profxj) [inline mhw_analysis/ai/ai_mhw_images.py:17]: this shouldn't be necessary.  You can and should
run

`python setup.py develop`

for each of the Repo's.

File: mhw_analysis/ai/ai_mhw_images.py [hunk 4]
```diff
@@ -0,0 +1,295 @@
...
+  """
+  Queries MHW_Events table given ranges (latitude, longitude, date), and returns
+  a dataframe containing all the entries of the table which meet that criteria. 
+
+  Query (SQL):
+  SELECT * FROM MHW_Events 
+  WHERE lat BETWEEN ? AND ? 
+  AND lon BETWEEN ? AND ?
+  AND date BETWEEN ? AND ?
+
```
Comment (profxj) [inline mhw_analysis/ai/ai_mhw_images.py:45]: Please add a comment for `connection`

File: mhw_analysis/ai/ai_mhw_images.py [hunk 5]
```diff
@@ -0,0 +1,295 @@
...
+  # get df
+  range_df = DataFrame(result.fetchall())
+  # range_df.columns = query.keys()
+
+  # return df
+  return range_df
+
+# used to save the tuples 
+# could include folder path as a param
```
Comment (profxj) [inline mhw_analysis/ai/ai_mhw_images.py:112]: Yes, please have a folder path.  

Best not to save anything into the Repo if we can avoid it.

File: mhw_analysis/ai/ai_mhw_images.py [hunk 6]
```diff
@@ -0,0 +1,295 @@
...
+	# create engine and check if it has the table we need
+	engine = sqlalchemy.create_engine('sqlite:///'+mhw_file)
+	engine.has_table("MHW_Events")
+
+	# connect to engine
+	connection = engine.connect()
+
+	# load table from metadata
+	metadata = sqlalchemy.MetaData()
+	mhw_tbl = sqlalchemy.Table('MHW_Events', metadata, autoload=True, autoload_with=engine)
```
Comment (profxj) [inline mhw_analysis/ai/ai_mhw_images.py:133]: Does this load the whole (10+Gb) table into memory?
If so, that must be very slow.  But maybe this is just
staging the table.  Shows my SQL ignorance.

File: mhw_analysis/ai/ai_mhw_images.py [hunk 7]
```diff
@@ -0,0 +1,295 @@
...
+	
+	# return connection and table
+	return connection, mhw_tbl
+
+def load_z500_noaa(path, date, any_sst):
+	# convert date to proper format
```
Comment (profxj) [inline mhw_analysis/ai/ai_mhw_images.py:139]: please add a doc string.

File: mhw_analysis/ai/ai_mhw_images.py [hunk 8]
```diff
@@ -0,0 +1,295 @@
...
+	ymd = date.year*10000 + date.month*100 + date.day
+
+	# create filepath
+	zfile = os.path.join(path, 'b.e11.B20TRC5CNBDRD.f09_g16.001.cam.h1.Z500.18500101-20051231-043.nc')
```
Comment (profxj) [inline mhw_analysis/ai/ai_mhw_images.py:143]: make the filename an optional parameter.

File: mhw_analysis/ai/ai_mhw_images.py [hunk 9]
```diff
@@ -0,0 +1,295 @@
...
+
+	# This is slow, so hold in memory
+	cubes = iris.load(zfile)
+
+	# get cube at ymd date
+	reload(cems_io)
```
Comment (profxj) [inline mhw_analysis/ai/ai_mhw_images.py:149]: you can remove this line.

File: mhw_analysis/ai/ai_mhw_images.py [hunk 10]
```diff
@@ -0,0 +1,295 @@
...
+	z500_cube = cems_io.load_z500(ymd, cubes)
+
+	# get z500 regridded data
+	z500_noaa = z500_cube.regrid(any_sst, iris.analysis.Linear())
+
+	return z500_noaa
+
+def get_noaa_extract(z500_noaa, lat, lon, width):
+	# create boundaries, make separate method when u start iterating over different frames
```
Comment (profxj) [inline mhw_analysis/ai/ai_mhw_images.py:158]: Add a doc string defining the arguments.  Or maybe I should! :)

File: mhw_analysis/ai/ai_mhw_images.py [hunk 11]
```diff
@@ -0,0 +1,295 @@
...
+
+def show_segmentation_map(lats, lons):
+	# ---MAKE PLOTS---
+
+	plt.clf()
+	ax = plt.gca()
+	fig = plt.figure()
+	ax1 = fig.add_subplot(111)
+
+	a = np.stack([np.arange(35.125, 44.875, 0.25) for _ in range(40)])
```
Comment (profxj) [inline mhw_analysis/ai/ai_mhw_images.py:181]: Is this just some hard-coded example?

File: mhw_analysis/ai/ai_mhw_images.py [hunk 12]
```diff
@@ -0,0 +1,295 @@
...
+	return cube.coord('longitude').points
+	
+def get_cube_data(cube):
+	return cube.data[:]
+
+# TODO: modularize lat/lon => row/col conversion, iterate lat/lon, speed up segmap stuff with spatial locality, re-write plots, multithreading(?)
+def main():
+
+	# prepare sql stuff, could make it a command line arg
+	mhw_file = '../../Downloads/mhws_allsky_defaults.db'
```
Comment (profxj) [inline mhw_analysis/ai/ai_mhw_images.py:216]: let's not hardcode the path.  You should use
an Environmental variable.

File: mhw_analysis/ai/ai_mhw_images.py (hunk 6 above)
Comment (Kamil-Kisielewicz) [reply mhw_analysis/ai/ai_mhw_images.py:133]: Just staging the table, my computer would be dying if it loaded the whole thing into memory

File: mhw_analysis/ai/ai_mhw_images.py (hunk 11 above)
Comment (Kamil-Kisielewicz) [reply mhw_analysis/ai/ai_mhw_images.py:181]: yeah that method needs to be revised but it's a task for when I can properly build the dataset.

**Pull request closed without merging**